File: backend/app/schemas/extraction.py
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class DocumentType(str, Enum):
    PDF = "pdf"
    DOCX = "docx"
    PPTX = "pptx"
    IMAGE = "image"
    UNKNOWN = "unknown"


class BlockType(str, Enum):
    HEADING = "heading"
    PARAGRAPH = "paragraph"
    LIST_ITEM = "list_item"
    TABLE = "table"
    CODE = "code"
    IMAGE = "image"


class ImageCategory(str, Enum):
    DECORATIVE = "decorative"
    CONTENT = "content"
    IMPORTANT = "important"
    UNKNOWN = "unknown"


class ImageClassificationResult(BaseModel):
    category: ImageCategory
    confidence: float
    score: float
    reasons: List[str] = Field(default_factory=list)


class ImageUrlRef(BaseModel):
    url: str


class ExtractedImageInfo(BaseModel):
    image_id: str
    filename: str
    mime_type: str
    hash_sha256: str
    width: Optional[int] = None
    height: Optional[int] = None
    size_bytes: Optional[int] = None
    saved_path: Optional[str] = None
    page_number: Optional[int] = None
    is_duplicate: bool = False
    duplicate_of: Optional[str] = None
    occurrence_count: int = 1
    preview: Optional[ImageUrlRef] = None
    asset: Optional[ImageUrlRef] = None
    classification: Optional[ImageClassificationResult] = None


class ExtractedBlock(BaseModel):
    block_type: BlockType
    text: Optional[str] = None
    level: Optional[int] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    image_info: Optional[ExtractedImageInfo] = None


class ExtractionResult(BaseModel):
    session_id: str
    document_type: DocumentType
    title: Optional[str] = None
    blocks: List[ExtractedBlock] = Field(default_factory=list)
    images: List[ExtractedImageInfo] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ExtractionRequest(BaseModel):
    session_id: str

File: frontend/src/types/extraction.ts
import type { ImageBlock } from "./image";

export type DocumentType = 'pdf' | 'docx' | 'pptx' | 'image' | 'unknown';

export type BlockType = 'heading' | 'paragraph' | 'list' | 'table' | 'code' | 'image';

export interface ListItemNode {
  text: string;
  level: number;
}

export interface BaseAstBlock {
  block_type: BlockType;
  metadata: Record<string, unknown>;
}

export interface HeadingAstBlock extends BaseAstBlock {
  block_type: 'heading';
  level: number; // 1-6
  text: string;
}

export interface ParagraphAstBlock extends BaseAstBlock {
  block_type: 'paragraph';
  text: string;
}

export interface ListAstBlock extends BaseAstBlock {
  block_type: 'list';
  ordered: boolean;
  items: ListItemNode[];
}

export interface TableAstBlock extends BaseAstBlock {
  block_type: 'table';
  headers: string[];
  rows: string[][];
}

export interface CodeAstBlock extends BaseAstBlock {
  block_type: 'code';
  language: string | null;
  code: string;
}

export interface ImageAstBlock extends BaseAstBlock {
  block_type: 'image';
  image_id: string;
  alt_text: string;
  asset_path: string;
  preview_url: string | null;
  ai_description: string | null;
}

export type AstBlock =
  | HeadingAstBlock
  | ParagraphAstBlock
  | ListAstBlock
  | TableAstBlock
  | CodeAstBlock
  | ImageAstBlock;

export interface DocumentAst {
  session_id: string;
  title: string | null;
  blocks: AstBlock[];
  images: ImageBlock[];
  metadata: Record<string, unknown>;
}